### ISSUE TITLE
`typing_extension.Self` infers the wrong type from `__new__`

### ISSUE THREAD

**Issue by Avasam** [microsoft/pyright#4668]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**
When using `__new__` directly (example below), the return type is the class itself instead of inferring from the first `classmethod` param (`cls`)

**To Reproduce**
```py
import enum
from typing_extensions import Self, reveal_type


class ActionLevel(enum.IntEnum):
    def __new__(cls, value: int, doc: str) -> Self:
        member = int.__new__(cls, value)
        reveal_type(member)
        member._value_ = value
        member.__doc__ = doc
        return member
```

**Expected behavior**
`member` should be `ActionLevel` not `int`

**Screenshots or Code**
![image](https://user-images.githubusercontent.com/1350584/220214648-472125fb-956f-4ea5-ab6f-feba45b403a7.png)

**VS Code extension or command-line**
Both.
`npx pyright@1.1.294 .\scripts\stubsabot.py`

**Additional context**
This was first found here: https://github.com/python/typeshed/pull/9694#discussion_r1109192407

**Comment by erictraut** [comment]
I'm not able to repro the problem you're seeing. Here's the behavior I see:

![image](https://user-images.githubusercontent.com/7040122/220215388-273b0e87-e4cf-4d86-8f8d-a51b7931f82a.png)

Can you confirm that when you paste the above code into a new editor window in a new project with default pyright settings that you get an error?

Perhaps you have an old version of `typing_extensions` or `typing_extensions-stubs` installed in your environment?

**Comment by Avasam** [comment]
`typing_extensions` is 4.5.0
I narrowed down what in the configs causes it: setting `"typeshedPath"`. In this case, my minimal configs are as such:
```json
{
    "$schema": "https://raw.githubusercontent.com/microsoft/pyright/main/packages/vscode-pyright/schemas/pyrightconfig.schema.json",
    "typeshedPath": ".",
}
```

Note this happens with `typing` too:
`npx pyright@1.1.294 .\scripts\stubsabot.py --pythonversion=3.11`
```
pyright 1.1.294
E:\Users\Avasam\Documents\Git\typeshed\scripts\stubsabot.py
  E:\Users\Avasam\Documents\Git\typeshed\scripts\stubsabot.py:8:21 - information: Type of "member" is "int"
  E:\Users\Avasam\Documents\Git\typeshed\scripts\stubsabot.py:9:16 - error: Cannot assign member "_value_" for type "int"
    Member "_value_" is unknown (reportGeneralTypeIssues)
  E:\Users\Avasam\Documents\Git\typeshed\scripts\stubsabot.py:11:16 - error: Expression of type "int" cannot be assigned to return type "Self@ActionLevel"
    "int" is incompatible with "ActionLevel" (reportGeneralTypeIssues)
2 errors, 0 warnings, 1 information
```

Is this something that can be fixes on typeshed's side?

I can confirm I have no `typing_extensions-stubs` in my env:
![image](https://user-images.githubusercontent.com/1350584/220217860-23d977c9-cad1-4a87-aa5a-a6b709f9613d.png)

**Comment by erictraut** [comment]
Ah, you're using a version of typeshed that is newer than what is bundled with pyright 1.1.294. That's an important detail! I'm able to repro the problem now, so I'll investigate further.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.295, which I just published. It will also be included in a future release of pylance.
